requests/maniphest_search.go

```diff
@@ -0,0 +1,116 @@
+	// Attachments specified what additional data should be returned with each result.
+	Attachments *SearchAttachments `json:"attachments,omitempty"`
+
+	*entities.Cursor
+	Request
+}
+
+// SearchAttachments contains fields that specify what additional data should be returned with sesarch results.
+type SearchAttachments struct {
+	// Subscribers if true instructs server to return subscribers list for each task.
+	Subscribers bool `json:"subscribers,omitempty"`
+	// Columns requests to get the workboard columns where an object appears.
+	Columns     bool `json:"columns,omitempty"`
+	// Projects requests to get information about projects.
+	Projects    bool `json:"projects,omitempty"`
+}
+
+// ManiphestRequestSearchOrder describers how results should be ordered.
+type ManiphestRequestSearchOrder struct {
+	// Builtin is the name of predefined order to use.
+	Builtin string
+	// Order is list of columns to use for sorting, e.g. ["color", "-name", "id"],
+	Order   []string
+}
+
+// UnmarshalJSON parses JSON  into an instand of ManiphestRequestSearchOrder type.
+func (o *ManiphestRequestSearchOrder) UnmarshalJSON(data []byte) error {
+	if o == nil {
+		return errors.New("maniphest search order is nil")
+	}
+	if jerr := json.Unmarshal(data, &o.Builtin); jerr == nil {
+		return nil
+	}
+
+	return json.Unmarshal(data, &o.Order)
+}
+
+// MarshalJSON creates JSON our of ManiphestRequestSearchOrder instance.
+func (o *ManiphestRequestSearchOrder) MarshalJSON() ([]byte, error) {
+	if o == nil {
+		return nil, errors.New("maniphest search order is nil")
+	}
+	if o.Builtin != "" {
+		return json.Marshal(o.Builtin)
+	}
+	if len(o.Order) > 0 {
+		return json.Marshal(o.Order)
+	}
+
+	return nil, nil
+}
+
+// SearchConstraints describes search criteria for request.
+type SearchConstraints struct {
+	// IDs - search for objects with specific IDs.
+	IDs            []string        `json:"ids,omitempty"`
+	// PHIDs - search for objects with specific PHIDs.
+	PHIDs          []string        `json:"phids,omitempty"`
+	// AssignedTo - search for tasks owned by a user from a list.
+	AssignedTo     []string        `json:"assigned,omitempty"`
+	// Authors - search for tasks with given authors.
+	Authors        []string        `json:"authorPHIDs,omitempty"`
+	// Statuses - search for tasks with given statuses.
+	Statuses       []string        `json:"statuses,omitempty"`
+	// Priorities - search for tasks with given priorities.
+	Priorities     []string        `json:"priorities,omitempty"`
+	// Subtypes - search for tasks with given subtypes.
+	Subtypes       []string        `json:"subtypes,omitempty"`
+	// Column PHIDs ??? - no doc on phab site.
+	ColumnPHIDs    []string        `json:"columnPHIDs,omitempty"`
+	// OpenParents - search for tasks that have parents in open state.
+	OpenParents    *bool           `json:"hasParents,omitempty"`
+	// OpenSubtasks - search for tasks that have child tasks in open state.
+	OpenSubtasks   *bool           `json:"hasSubtasks,omitempty"`
```

Comment:
Should this and OpenParents be just `bool`?

Reply:
if it is just bool then the value gets into request and messes it. The reason for pointer here is to be able to use omitempty so that OpenParent parameter is not in request and is not used by the server.

Reply:
With omitempty, a value will not be sent if it's false.

https://play.golang.org/p/QfHtNxtGcQ6

That should be fine because the server defaults both those fields to false anyway.

By the way, according to [the source](https://github.com/phacility/phabricator/blob/53b9acfb7d58ac5f33424b84df35757a5c68aba1/src/applications/maniphest/query/ManiphestTaskSearchEngine.php#L99), this is not to *also* show open parents or subtasks but *only* tasks with open parents or subtasks.

The documentation here should probably include the "only".

Reply:
@abhinav indeed false is not sent with omitempty and it works as expected. However, bool w/o omitempty returns very few results so I don't see it that server defaults to false. It looks more like tri-state. Maybe that's why I put put a pointer there originally, don't quite remember at this point.